### ISSUE TITLE
Parameterless prepared select throws (code 101)

### ISSUE THREAD

**Issue by pti** [simolus3/sqlite3.dart#145]
Hi,

If I use a prepared select statement for the second time, I get an SqliteException with code 101. For example with the simplified example code below the first select works okay, but the second `ps.select()` throws the exception shown below.

```
import 'package:sqlite3/sqlite3.dart';

void main() {
  final db = sqlite3.openInMemory();
  db.execute('CREATE TABLE IF NOT EXISTS foo (id INTEGER PRIMARY KEY, name TEXT)');
  final ps = db.prepare('SELECT * FROM foo');
  final result1 = ps.select();
  final result2 = ps.select();
}

---

Unhandled exception:
SqliteException(21): no more rows available, no more rows available (code 101)
  Causing statement: SELECT * FROM foo
#0      throwException (package:sqlite3/src/ffi/impl/exception.dart:40:3)
#1      PreparedStatementImpl._selectResults (package:sqlite3/src/ffi/impl/statement.dart:154:7)
#2      PreparedStatementImpl.select (package:sqlite3/src/ffi/impl/statement.dart:128:12)
#3      main (file:///.../bin/main.dart:8:22)
#4      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:297:19)
#5      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:192:26)
```

If I add a parameter to the query, e.g. `SELECT * FROM foo WHERE id > ?`, then both select-calls return without an exception getting thrown. Also, if I add `_finalizable._variablesBound = true;` to `lib/src/ffi/impl/statement.dart` line 190's if-block, before returning then the example works too (no idea if that is the cause, just tried something).

sqlite3 version 1.9.2.

**Comment by simolus3** [comment]
This is definitely a bug, thanks!

I am struggling to understand how an exception could be thrown though. If you look at the check, error 101 (`SQLITE_DONE`) should be handled there:

https://github.com/simolus3/sqlite3.dart/blob/a8e99891f65c29bf300add8d9d7a234cff6f256d/sqlite3/lib/src/ffi/impl/statement.dart#L158-L160

I then thought that we might be missing data by forgetting to call `sqlite3_reset`, but it seems like (at least my) sqlite3 implicitly resets a statement after `_step()` has returned `SQLITE_DONE`.

So while this should be fixed in the library, I'm wondering which sqlite3 (the native library, not the Dart package) you are using? Something from your OS or a custom build? What does `sqlite3.version` return?

**Comment by simolus3** [comment]
Ah well, the test indicates the problem - looks like the sqlite3 on macOS is behaving differently.

**Comment by pti** [comment]
Yep, I was using macOS - `print(sqlite3.version)` outputs:
> Version(libVersion: 3.39.5, sourceId: 2022-10-14 20:58:05 554764a6e721fab307c63a4f98cd958c8428a5d9d8edfde951858d6fd02daapl, number: 3039005)

On Linux the exception isn't thrown, version:
> Version(libVersion: 3.37.2, sourceId: 2022-01-06 13:25:41 872ba256cbf61d9290b571c0e6d82a20c224ca3ad82971edc46b29818d5dalt1, number: 3037002)

**Comment by simolus3** [comment]
Fixed in `sqlite3: 1.9.3` - thanks again for the report!

**Comment by pti** [comment]
That was quick! 👍
